### ISSUE TITLE
Onsite

### ISSUE THREAD

**Issue by krujeen** [krujeen/Training_KMITL#3]
### การ Train Model Yolov5 ด้วย Colab
**Download yolov5 ลงใน Colab**
[https://github.com/syanyong/ml_yolov5_training](https://github.com/syanyong/ml_yolov5_training)

**เตรียม DataSet สำหรับ training**
[https://app.roboflow.com/](https://app.roboflow.com/)

**Upload ไฟล์ที่ได้จาก ROBOFLOW**
**Unzip ไฟล์ที่ได้จาก ROBOFLOW**

**แก้ไฟล์ data.yaml**
```
train: ./train/images
val: ./valid/images
test: ./test/images
```

**เชื่อมต่อ wandb**
[https://wandb.ai/](https://wandb.ai/)

**สั่งให้ train model**

>ไฟล์ /content/yolov5/runs/train/exp/weights/best.pt  คือ model ที่ได้จากการ train เพื่อนำไปใช้

**Comment by krujeen** [comment]
### การใช้ Model Yolov5 ด้วย  VS Code

**Setup Python**
[https://www.python.org/](https://www.python.org/)

**VS Code**
[https://code.visualstudio.com/](https://code.visualstudio.com/)

**เชื่อมต่อ VS Code กับ Python**
เลือก Extensions
ค้นหา Python แล้ว install


**Download yolov5**
[https://github.com/ultralytics/yolov5](https://github.com/ultralytics/yolov5)


**สร้าง Virtual Environment**
(mac)
```
% python3 -m venv venv  
% source ./venv/bin/activate
```
(window)
``` 
% python -m venv venv  
% venv\Scripts\activate
```
_[วิธีแก้ไข เมื่อ window สร้าง Virtual Env ไม่ได้](https://syanyong.notion.site/Fixed-virtualenv-won-t-activate-on-Windows-d738f939a62244c5bc94c13941696c3f)_

**ติดตั้ง library ที่จำเป็น**
```
% pip install -r requirements.txt
```

**Comment by krujeen** [comment]
**เริ่มทดสอบ**
```
% python3 detect.py
```

เรียก resource ที่เกี่ยวข้อง
> detect: weights=yolov5s.pt, source=data/images, data=data/coco128.yaml, imgsz=[640, 640], conf_thres=0.25, iou_thres=0.45, max_det=1000, device=, view_img=False, save_txt=False, save_conf=False, save_crop=False, nosave=False, classes=None, agnostic_nms=False, augment=False, visualize=False, update=False, project=runs/detect, name=exp, exist_ok=False, line_thickness=3, hide_labels=False, hide_conf=False, half=False, dnn=False, vid_stride=1

บอกเวอร์ชั่น YOLO Python  และ CPU ที่ใช้
> YOLOv5 🚀 2023-5-3 Python-3.11.3 torch-2.0.0 CPU

Download dataset จาก COCO
> Downloading https://github.com/ultralytics/yolov5/releases/download/v7.0/yolov5s.pt to yolov5s.pt...

ผลลัพธ์ที่ได้จากการ run model ประกอบด้วย จำนวน layer, ภาพที่ใช้, เวลา และโฟลเดอร์ที่เก็บไฟล์ผลลัพธ์
> YOLOv5s summary: 213 layers, 7225885 parameters, 0 gradients
image 1/2 /Users/krujeen/programming/yolov5-master/data/images/bus.jpg: 640x480 4 persons, 1 bus, 86.4ms
image 2/2 /Users/krujeen/programming/yolov5-master/data/images/zidane.jpg: 384x640 2 persons, 2 ties, 67.5ms
Speed: 0.4ms pre-process, 76.9ms inference, 1.1ms NMS per image at shape (1, 3, 640, 640)
Results saved to runs/detect/exp

**Comment by krujeen** [comment]
**Free Dataset**
[https://cocodataset.org/](https://cocodataset.org/)

**Comment by krujeen** [comment]
**ใช้งานกล้อง** 
% python3 detect.py --source 0

**Comment by krujeen** [comment]
**ใช้งานกล้อง** 
```
% python3 detect.py --source 0
 ```

**ใช้งานกล้อง IP CCTV** 
```
% python3 detect.py --source 'rtsp://admin:password@192.168.1.2:554/cam/realmonitor?channel=1&subtype=1'
```

**ต้องการใช้ model ที่ train มาใช้** 
> ให้ copy ไฟล์ best.pt มาใส่ในโฟลเดอร์ yolov5

```
% python3 detect.py --weights best.pt  --source 'rtsp://admin:Otmadmin1234@10.10.9.115:554/cam/realmonitor?channel=1&subtype=0'
```

> subtype=0 คือสตรีมหลัก
> subtype=1  คือสตรีมย่อย

**keyword ในการค้นหา**
> tp link tapo rtsp

วิธีออกจากกล้อง ให้กด Ctrl+C

**Comment by krujeen** [comment]
**เขียนคำสั่งเพื่อใช้เปิด-ปิดกล้อง**
[https://github.com/krujeen/Training_KMITL/blob/main/run1.py](https://github.com/krujeen/Training_KMITL/blob/main/run1.py)
```
import torch
import cv2
import numpy as np

# vid = cv2.VideoCapture(0)
vid = cv2.VideoCapture("rtsp://admin:passwordd@192.168.1.2:554/stream1")  #แก้ไขให้ตรงกับกล้อง ip CCTV

while(True):
    ret, frame = vid.read()
    cv2.imshow('YOLO', frame)
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

vid.release()
cv2.destroyAllWindows()
```

**load model**
ก่อน Video Capture
```
model = torch.hub.load('ultralytics/yolov5', 'yolov5s', device='cpu')
```

ถ้าต่อเน็ตไม่ได้ให้ใช้คำสั่งนี้แทน
```
model = torch.hub.load('.', 'custom', path='yolov5s.pt', source='local')
```

ก่อน imshow
```
    results = model(frame)
    results.print()
```


**Plot bounding box from Pytorch Hub**
[https://gist.github.com/syanyong/5fd83ff7d006d4566e115f9dbf203905
](https://gist.github.com/syanyong/5fd83ff7d006d4566e115f9dbf203905)
ใส่ก่อน while(True)

```
# Plot bounding box from Pytorch Hub
def plot_boxes(result_dict, frame):
    for ob in result_dict:
        rec_start = (int(ob['xmin']), int(ob['ymin']))
        rec_end = (int(ob['xmax']), int(ob['ymax']))
        color = (255, 0, 0)
        thickness = 3

        cv2.rectangle(frame, rec_start, rec_end, color, thickness)
        cv2.putText(frame, "%s %0.2f" % (ob['name'], ob['confidence']), rec_start, cv2.FONT_HERSHEY_DUPLEX, 2, color, thickness)
    return frame
```


เพิ่มหลัง import numpy 
```
import json
```

เพิ่มหลัง results = model(frame)
```
    result_jsons = results.pandas().xyxy[0].to_json(orient="records")
    result_dict = json.loads(result_jsons)
    frame2 = plot_boxes(result_dict, frame)
```

**แก้ไข เพื่อใช้ model ของตนเอง**
```
model = torch.hub.load('.', 'custom', path='best.pt', source='local')
```

**Comment by krujeen** [comment]
### การแจ้งเตือนผ่าน Line
[https://notify-bot.line.me/th/](https://notify-bot.line.me/th/)

1. เลือก MyPage
2. คลิก Generate  token
3. Please enter a token name to be displayed before each notification.  ตั้งชื่อการแจ้งเตือน
4. คลิกสร้าง Generate  token
5. คัดลอก Your token 
6. ติดตั้ง line_notify ใน terminal ของ VS Code
```
% pip install line_notify
```

7. สร้างไฟล์ line.py
```
from line_notify import LineNotify

ACCESS_TOKEN = "Your token"
notify = LineNotify(ACCESS_TOKEN)
notify.send("แจ้งเตือน")
```

8. สร้างฟังก์ชัน sendText
```
def sendText(text):
    notify.send(text) 
```

9. import ฟังก์ชัน sendText  ในไฟล์ run1.py
เพิ่มหลัง import json
```
import line
```
10. เพิ่มคำสั่ง เรียกใช้ฟังก์ชัน sendText ในไฟล์ run1.py
```
line.sendText("Detected")
```
 
11. ส่งรูปภาพ ให้เพิ่มฟังก์ชัน sendImage ใน line.py
```
def sendImage(text,path):
    notify.send(text,image_path=path)
```

12. เพิ่มคำสั่ง เรียกใช้ฟังก์ชัน sendImage ในไฟล์ run1.py
```
   cv2.imwrite("./output.jpg",frame)
   line.sendImage("Detected","./output.jpg")
```

**Comment by krujeen** [comment]
ref.
[https://github.com/syanyong/yolov5-PyTorch-Hub-chapter](https://github.com/syanyong/yolov5-PyTorch-Hub-chapter)
